[READY] Ripley Rework

zxaber opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Changes the Ripley to be a light-weight mech with little protection, but faster speed. The old version is available with a conversion kit.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
The Ripley currently has no real purpose, except as a bulky EVA suit. The Firefighter does everything the Ripley can do but better, and with the nerf to mech plasma cutters, the Ripley can't really stand up as a mining vehicle if any fauna are around. These changes should help guide it into a standard station role, at the cost of making it worse during emergencies. 
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: Zxaber
tweak: Ripley now has an open cabin. This means it cannot protect against atmospheric dangers, and it does not provide the user cover from ranged attacks, but moves quicker and is slightly cheaper to manufacture.
add: A conversion kit is available in the exosuit fabricator that gives the Ripley a pressurized cabin as before, but the added weight will slow it down again.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

In more detail;
Ripley has been given a wire screen, which quite visually hints that it does not have a pressurized cabin. The internal tank is not present, nor is the HUD button for enabling it. The wire screens do little to block ranged attacks either, happily passing both the effects and damage right on to whoever is inside. If the Ripley is not occupied, or if an AI or MMI is the pilot, the mech itself will take the damage.

On the other end, the Ripley no longer takes Plasteel to create; That step has been replaced with 10 metal rods to make the screen. The open cabin also allows a user to enter quicker (currently taking a quarter of the time of any other mech). Finally, the lost weight of ditching the enclosed cabin allows the Ripley to move quicker (currently at 2.5 in standard atmospheric conditions and 1.5 in low atmos).

The pilot is a visible silhouette, and as such examining the Ripley will let you know who is inside.
![image](https://user-images.githubusercontent.com/37497534/52921235-d41d0e00-32c9-11e9-8bf7-1f70dc1913dd.png)
I specifically made it so that line doesn't show to the operator if they examine their own mech (Since "You can see \<your own name\> inside" seemed weird).
On the other hand, the silhouette doesn't show if the mech is controlled by an AI or MMI.
![image](https://user-images.githubusercontent.com/37497534/52921293-79d07d00-32ca-11e9-84f1-94a3956d1b12.png)
The Ripley also no longer blocks vision, since it has a window and all.

Both the Firefighter, and the Death-Ripley are unchanged.

![image](https://user-images.githubusercontent.com/37497534/52921418-b9e42f80-32cb-11e9-826f-6135f479a9c8.png)

Due to feedback, the old Ripley can be created by using a conversion kit on the new Ripley. Thematically, it's a MK-I to MK-II conversion. MK-II (the old Ripley) keeps its old speed but gains the atmospheric protection back, and is entirely equal to the existing one. As such, if your goal from the start is a pressurized utility mech, the Firefighter is a better choice. But if you already have a MK-I Ripley and suddenly _need_ to space-walk or something, you can convert it to MK-II. The conversion cannot be undone.

zxaber committed:
Bunches of changes

Indeed, a whole barrel full

zxaber committed:
Update ripley.dm

zxaber committed:
Update mecha.dm

fluffe9911 commented:
I mean sounds cool but I dont exactly see how this will make people consider choosing this over the firefighter since this seems to just make the mech even more useless then it was before

zxaber committed:
Update ripley.dm

wesoda25 commented:
If you really wanted to be daring you’d remove mechs all together....

ATH1909 commented:
> I mean sounds cool but I dont exactly see how this will make people consider choosing this over the firefighter since this seems to just make the mech even more useless then it was before since now it isnt even space proof

Because now brainlet roboticists who can't figure out how to get plasteel can start blowing huge amounts of metal on a mech without leaving their department! And who really needs borg shells, anyway? That MMI'd brain on the ground can just, I dunno, chat on comms on something while I make my mech. Actually, I'm pretty sure that it's braindead- that's what the buzzing noise when I put it in the MMI and its turned-off lights mean, right?

/joke

In all seriousness, though, I like this change. If nothing else, it makes the Ripley closer to the Ripley from the movies. I also like the idea of having a fast, cheap mech- seems perfect for posibrains that want to be mechs instead of borgs. You can also mitigate the "doesn't protect you from projectile attacks" drawback of the mech by either having a silicon pilot it or wearing an HoS hardsuit and having dermal refractive surface nanites active (if you're augged too, that might even be an upside, making your mech much more resistant to laser attacks than it otherwise would be (since they'll hit you and deal no damage instead)).

.Maybe you could make it so that projectiles that are aimed at the Ripley can hit the pilot, but projectiles that aren't aimed at the Ripley (but still hit it) hit the Ripley instead (of its pilot)?

Caiggas commented:
I like this. It changes the current riply from a weaker firefighter to a fundamentally different niche. Light/fast cargo loader/construction mech. Are there already cargo expansion upgrades to go with the loading module? If not, it would make a good addition. Maybe have projectiles have a percentage chance to hit occupant or mech, like 50/50 or whatever. Just a thought. All in all, this is a good piece of work.

zxaber commented:
So far, there's no expansion module, but the Ripley can hold fifteen objects (including crates/lockers) by default. A module that extends it further might be cool.

I did consider having a chance for shots to ping off the mech and deal damage to it (probably something closer to 15-25%), but I'm not sure if adding RNG makes things more fun.

habader commented:
maybe you should make it cost less metal overall too

PKPenguin321 commented:
Not sure why this replaces the RIPLEY instead of just being a new thing

zxaber commented:
Mostly because the existing Ripley has no purpose. It's too slow to use on station, it's not fireproof, you get lost in space if you try to leave the station, and you shouldn't mine with it because you'll end up trapped against lava and surrounded by things you can't kill. The whole reason for the rework is to give it an identity and some sort of use.

nero1024 commented:
Because the current Ripley is essentially obsolete due to the presence of the Firefighter Ripley.

PKPenguin321 commented:
So? There's still nothing to lose if we keep it in the code.

PKPenguin321 commented:
Also that sounds more like an issue with firefighters being unbalanced than with RIPLEYs

zxaber commented:
What do we lose by changing it to this version?

zxaber commented:
@habader
> maybe you should make it cost less metal overall too

So far, the Ripley is at a total of 60 sheets of metal; 50 for the initial printing, 5 for the outer plating, and 10 rods for the wire screen.

I could lower the cost of the chassis a bit, but it's at 10 sheets as it is, and it would feel weird dropping that too heavily. I can't easily adjust the other parts, because they're shared by the Firefighter and the changes would bleed over. Anyway, to be honest, I think sixty metal for the new Ripley is fair, considering there's no longer a need for plasteel.

PKPenguin321 commented:
>What do we lose by changing it to this version?

The option of having a slower, space-proof RIPLEY

zxaber commented:
You could just raid EVA for a suit. The new Ripley doesn't prevent you from wearing a space suit, it just doesn't provide the protection on its own.

Or, y;know, make a Firefighter. Yeah, it requires 15 plasteel, but you start with at least 10 plasteel in Robotics, and you can just take apart the tables in the experimentor room for the last five. Or ask Engineering. Or wait until Mining brings back plasma and print as much plasteel as you need from the protolathe in R&D.

Are we really better off having two almost identical mechs with no real difference except one is not fireproof and costs just a tiny bit lest in mats?

davethwave commented:
I think this could end well. Maybe could add cargo expansion upgrades and the like for crates and the like. Don't forget that buying metal isn't too expensive if need be.

MMMiracles commented:
Just make it a split path mid-way through production, where you can either choose to go with the lightweight-but-open design or heavier-but-spaceproof design.

tralezab commented:
can't we just have a rugged scrap mech with this wired front idea? if anything, the ripley should be the only spaceproof because of it's role as the support/engineering/mining mech

Chronitonity commented:
If you're concerned about "losing" the old Ripley then why not just add in a module that replaces the wire screen with a pressurized Plasteel cabin and makes it function exactly like the old Ripley?

PKPenguin321 commented:
>Are we really better off having two almost identical mechs with no real difference except one is not fireproof and costs just a tiny bit lest in mats?

Unambiguously yes because there's no downside to this?

Tlaltecuhtli commented:
why use ripley then?

MACIEKBAKI commented:
> If you're concerned about "losing" the old Ripley then why not just add in a module that replaces the wire screen with a pressurized Plasteel cabin and makes it function exactly like the old Ripley?

That seems like a good idea, maybe when you put in the wire screen make it so you can use plasteel instead for the more armored, slower ripley?

Nickkoranda commented:
>That seems like a good idea, maybe when you put in the wire screen make it so you can use plasteel instead for the more armored, slower ripley?

This exists, it's called the firefighter. Ripley has no special use right now than newbie roboticists making it because they don't know about the firefighter.

81Denton commented:
Another reason why nobody builds Ripleys is also because mech-mounted tools plain suck compared to standard or even advanced tools like the jaws of life.
It's way out of scope for this PR, but I feel like more people would build them if for example, mech clamps let you pry airlocks open.

necromanceranne commented:
Making a firefighter was so bloody simple and quick that I just never considered wasting mats on anything else. Ripleys were just always lagging behind but Firefighters themselves were not really the problem.

A faster mech is a good one for simple construction and supply use, and one that's cheap is also quite handy too. This especially goes for lavaland, even without the lavawalking. I could easily see having a Ripley be useful for cargo in other ways, especially since you can hold so many crates and stack them into the same tile by unloading everything. If they happened to just get every crate on the station that can be hard to fit into the cargo shuttle all at once. And collecting all of it something of a chore since you have to go back and forth to gather everything.

Since the mech is so cheap, maybe consider letting cargo order in a deconstructed ripley to put together for a fairly reasonably price. Something like 4k, which is their starting budget. It'd make for a nice alternative to the usual roundstart of insuls if they think they can use it to scoop up all the station's crates and fuel tanks fast with it.

datorangebottle commented:
> Since the mech is so cheap, maybe consider letting cargo order in a deconstructed ripley to put together for a fairly reasonably price. Something like 4k, which is their starting budget. It'd make for a nice alternative to the usual roundstart of insuls if they think they can use it to scoop up all the station's crates and fuel tanks fast with it.

I dislike this because it removes something the Roboticist can build roundstart.

The new ripley has to be fast enough that the space protection is actually a tradeoff instead of a huge downside.

zxaber commented:
Well, I don't know if it would impact too badly with robotics; engineering can either build or buy emitters, for example. Honestly, if Economy is ever going to mean anything, we need to have things people actually care about in the cargo catalog. Since the Ripley is supposed to be a pre-existing mech (no research is required), it would make a good candidate. Still, that's a change for another PR.

The new Ripley's move delay is set to 2.5 it normal atmos conditions. Lower is faster, players are currently at 1.5 without speed debuffs and the old Ripley and current Firefighter are at 4. I'll make a video demonstrating the new speed (and the entry speed), but it feels only slightly slower than walking. If the new Ripley is ever in a low atmos environment (presumably the pilot grabbed a suit), it will move at player speed.

zxaber commented:
Also, to the ideas of having branching build paths for new and old Ripley from the same chassis;

The build paths are currently locked into being linear. The Firefighter uses a new chassis rather than branching out after a fire suit is added, even though all the other parts are Ripley parts. Making a branch would involve even more snowflaky code and I'm not sure it would be worth it. If I _do_ add a way to get current Ripley, I'll probably name it "Ripley Prototype MK-II" , to be thematically a prototype being tested before the Firefighter was developed, and make a separate chassis for it.

PKPenguin321 commented:
>Another reason why nobody builds Ripleys is also because mech-mounted tools plain suck compared to standard or even advanced tools like the jaws of life.

??? literally, LITERALLY one or two days ago i rolled robo specifically to build a RIPLEY and stomp around all mech-like because i enjoy doing that
and again, if RIPLEYs are useless because of firefighters, how is that a RIPLEY problem and not a firefighter problem???

zxaber commented:
The thing is, the part that makes Firefighters so much better is that they're fire-proof.

What purpose does the Ripley serve, and what purposes has it served in the past?
- Cargo Forklift? Well, it's rather slow compared to people just pulling boxes around. It can hold a bunch, bit it's a bit of a pain walking all over the station at a run_delay of 4.
- Space Exploring? Mechs can't use jetpacks, and so once you leave you're committed to just drifting. Even if you bring a cutter for the recoil, you can't turn the mech without some purchase, so you at best can go back to your starting location.
- Mining? Not with the cutter nerf. Now that the utility mechs cannot fight Lavaland fauna, Ripley brings absolutely no benefit to just mining on foot. Humans move faster and can actually kill anything that threatens them. Firefighter's have the advantage in that they can cross lava, but that's it.
- Emergency Response? Maybe a bit, but the mech RCD can't make doors, and you're unable to really do anything else. The wire layer kinda sucks hard, and so rebuilding in general is better off on foot, in a EVA suit, with a handheld RCD. Plus, again, if there's any spot missing floor altogether, the mech will get stuck. You can attach an extinguisher for fighting fires, but the Ripley is not fireproof and you'd be better off making a Firefighter.
- Safety Bunker? Perhaps, but the combat mechs are better anyway for that use, and Durands aren't like super hard to make.
- Walking around like you're king of the world because you can put six parts together and make a yellow mech? Alright, sure. Hopefully no one needed the robo to be at their workshop for anything. (And anyway, Durands are better suited for this purpose.)

Most of the purpose of this change is that the Ripley is better suited to what the name implies; Autonomous Power Loader Unit. As in, loads boxes. It hopefully fills the role of a forklift, since it's faster than it use to be, while providing downsides to that speed so that it's not just a flat buff. It is not designed for use in emergency situations, and any rolls that were filled in that way can just be done with the Firefighter. And anyway, you can just use an EVA suit to bypass the atmos thing (though not the ranged damage thing), most stations have like eight suits at least.

imsxz commented:
There’s no point to the yellow Ripley at the moment when it’s just a downgrade to firefighter in every way. Nothing of value lost even if this was a PR to simply remove the lesser Ripley altogether, this’ll hopefully make firefighter not the only budget mecha we see made

zxaber committed:
Glad I'm testing this

zxaber committed:
herpaderp

zxaber commented:
Also, like I promised earlier, video comparing movement speeds of Firefighter and new Ripley.
https://streamable.com/1sqad

BuffEngineering commented:
I really like this since it makes them closer to what they're based off of while filling a new niche, but after looking at the video and trying it out myself I think the speed of 2.5 is still a little slow. For cargo to get one they either have to bug robotics or order/steal the components themselves, which takes time and is less efficient than just buying the techs some nuka cola to hurry the process of gathering sellable stuff. Giving them a speed of something like 2.2 or even 2.0 in atmosphere would still allow people to run away or kite them with a gun and keeps them slower than or on par with the odysseus. 

Also am I missing something or are the code comments for fast_pressure_step_in and slow_pressure_step_in reversed?

zxaber commented:
I wouldn't mind dropping the delay to 2.0. I'm trying not to go too far overboard, but 2 should be fine. Additionally, the more I think about it, the more I like Cargo being able to order parts for a Ripley. If I do add that, it'll be in a separate PR (and I'd probably want to remove the Ripley bounty).

I dunno about the comments, I guess they are backwards. I didn't even really pay them much mind.

OverlyGeneric commented:
>nobody questioning that the increased movement speed makes the Ripley more robust for megafauna killing. 🤔

zxaber commented:
Cutter damaged got super-nerfed. If you want to spend two-minutes per goliath, then go ahead, but the increased speed isn't really going to make it go quicker.

As for megafauna, then eh? It's not fireproof, nor bulletproof, so Colossus and Drake will get their damage in. It'll hold up against Bubblegum and the purple disco stick, but again, you're doing like minimum damage, and only a few missteps will make the mech useless due to some malfunction or other. I really don't think it would be worth the trouble.

zxaber committed:
Ripley MK-I to MK-II upgrade

text goes here. Also fixed the pressure comments for step_in.

zxaber commented:
Ripley can now be upgraded from the New version (MK-I), to the old version (MK-II). Yeah, it's confusing but the idea is that the MK-II was a prototype being tested before the Firefighter (MK-III) was developed.

Anyway;
![image](https://user-images.githubusercontent.com/37497534/52999216-0e75d080-33da-11e9-8ddc-baa5efad700e.png)
One-way upgrade, because it actually just makes a new MK-II, moves all the gear over, and then deletes the MK-I. So it requires maintenance procedures to be enabled so that the owner has some control (since the MK-II _is_ slower by far than the MK-I). It also requires the mech to not have a pilot, because I really didn't want to deal with the mess of moving a player too.

Printable from the Exo-fab, on the same node as the Ripley itself (so no research needed), and costs five metal and five plasma (equal to the five plasteel that current Ripley costs). You can just stick five of Robo's round-start plasteel in to a fab, it'll work.

ExcessiveUseOfCobblestone commented:
> Ripley's are bad
>> here's a PR that makes them good imo
>>> I also added an "upgrade" that reverts my what I believe to be good changes

lol okay...

zxaber commented:
It was heavily requested of me to have the old model still available, so ¯\\_(ツ)_/¯

ExcessiveUseOfCobblestone commented:
My take on it is that by having the old style as an upgrade option you're implying your changes makes the ripley **worse** at base so I don't see why I would merge a nerf on something that is already unused.

zxaber commented:
It's thematically an upgrade because I wanted to sorta have an implied design path; MK-II was toying with the idea of a pressurized cabin, before it was scrapped for the MK-III Firefighter.

Ultimately, I think the MK-I will be more useful in more situations, there was just someone very passionate about old Ripley existing still, and I figured an item to transform I into II was better than a whole other build path.

Caiggas commented:
> It's thematically an upgrade because I wanted to sorta have an implied design path; MK-II was toying with the idea of a pressurized cabin, before it was scrapped for the MK-III Firefighter.
> 
> Ultimately, I think the MK-I will be more useful in more situations, there was just someone very passionate about old Ripley existing still, and I figured an item to transform I into II was better than a whole other build path.

Unfortunately you can't make everyone happy. I'd say you found the best middle ground. _Now we just need a PR to rework the firefighter into an ACTUALLY USEFUL firefighter._

zxaber commented:
Denton's idea of the clamp opening firelocks would probably help, and I might make a PR for that next.

zxaber commented:
Right, so a few new things;
The MK-I to MK-II upgrade is now described as a conversion kit rather than a flat upgrade, and both the item and resulting mech's descriptions indicate that a trade off of speed for atmospheric protection is made. The MK-II is still a worse mech than the MK-III Firefighter, and doesn't really fill a niche like the MK-I, but if you have a MK-I around and suddenly have a dire need for a space-capable mech, the conversion is cheaper than a new Firefighter. Hopefully this gets at what my intentions were as to the MK-II existing.

MK-I's cargo hold needs to be empty before the conversion now. The processes also makes a wrench sound.

Currently trying to figure out a way to ignite the pilot of a MK-I if the surrounding turf is burning. I _think_ that's the last thing I need for this PR.

zxaber committed:
Various things.

zxaber committed:
First pass at fire

also, sanity checks and reorganizations

zxaber commented:
So, MK-I now sets the operator on fire if the mech is standing in fire. Thing is, the fire code is a bit complicated and involved fire stacks, and I couldn't figure out how many stacks normal fires give to a human. So each process cycle it adds one stack to the operator, until they have 20.  It's not like _super_ quick, but it does fire at least once a second (I think - my testing is currently being done over Teamviewer).

Anyway, I'd _love_ some feedback on if this is the correct way to apply fire to an unlucky operator.

Militaires commented:
@ExcessiveUseOfCobblestone >turns the ripley from somewhat kinda bad to total trash
>pr that makes them good

![image](https://user-images.githubusercontent.com/16103879/53454469-fd5b3e00-3a2f-11e9-852d-72453d640355.png)

zxaber commented:
Old Ripley had some advantages, but if you were making one, you might as well make a firefighter. The point of this PR was to give it a niche to seperate it from being just a weaker Firefighter.

Anyway, had some real life stuff come up; I want to adjust the fire stacks and fix a description, and then it will be ready for review probably in a day or two.

Militaires commented:
fair enough

zxaber committed:
Drop fire stacks to 5, max

ExcessiveUseOfCobblestone commented:
@Militaires I was speaking as if I was the PR author.

The PR desc describes the changes as an improvement to the original ripley, yet also kept in a (implied worse) version in. Without justification (which was provided), it seemed pointless to bother with this PR.

zxaber committed:
bug fixes

zxaber commented:
Using a MK-I in a fire now only gives fire_stacks up to five. Far as I can tell, fire_stacks only affects the duration of being on fire, not the damage, so you'll still be just as on fire and it's generally not a good idea to dive in a fire with the mech unless you have an atmos hardsuit. Information about how fire_stacks are applied (in terms of number) is proving to be illusive, and I don't want to add more stacks than a fire would give without the mech, so I'll leave them at five. That is at just barely being able to remove the fire with one resist, if I read the code right.

Fixed some bugs. I love to death the concealed weapon bay, but why it was not just a variable flipped on the mech when attached, I have no idea.

I can't find anything else not working, so I think I can remove the WIP tag now.

zxaber committed:
A few contruction sprite tweaks

AnturK commented on code/game/mecha/mecha.dm:
```diff
@@ -205,6 +213,11 @@
+/obj/mecha/bullet_act(obj/item/projectile/P)
+	if (enclosed || !occupant || silicon_pilot)
+		return ..()
+	occupant.bullet_act(P) //If the sides are open, the occupant can be hit
```
return value here ?

zxaber replied:
This return stops the proc and makes the mech take damage as normally. It only passes the damage on to the pilot if one exists, as a non-silicon, and the mech has an open canopy.

AnturK replied:
I mean return value from hitting the pilot needs to be passed back to projectile.

zxaber replied:
Is that not handled by the occupant's bullet_act function?

AnturK replied:
No, the caller won't get any information about penetration/blocking etc here. You need to return it.

zxaber replied:
Alright, I _think_ I read how bullet_act is handled correctly, and have given a return BULLET_ACT_HIT to the function. I've also moved it to mecha_defense, where it should have been before anyway.

AnturK commented on code/game/mecha/mecha.dm:
```diff
@@ -400,6 +417,14 @@
 	diag_hud_set_mechtracking()
 
+	//Check if we should ignite the pilot of an open-canopy mech
+	var/turf/target = get_turf(src)
```
This should be in fire_act or similar instead of process.

zxaber replied:
The fire_act() proc isn't available for mechs - it's a mob-only proc, so I can't use it to check for fire.

AnturK replied:
fire_act is atom proc.

zxaber replied:
Huh, dunno what I did wrong last time I tried it, but I got it working now. Will push an update out soon.

AnturK commented on code/game/mecha/mecha_topic.dm:
```diff
@@ -71,19 +71,24 @@
-						<b>Airtank temperature: </b>[tank_temperature]&deg;K|[tank_temperature - T0C]&deg;C<br>
-						<b>Cabin pressure: </b>[cabin_pressure>WARNING_HIGH_PRESSURE ? "<span class='danger'>[cabin_pressure]</span>": cabin_pressure]kPa<br>
-						<b>Cabin temperature: </b> [return_temperature()]&deg;K|[return_temperature() - T0C]&deg;C<br>
+						<b>Air source: </b>[internal_tank?"[use_internal_tank?"Internal Airtank":"Environment"]":"Enviroment"]<br>
```
typo, also you could probably display air pressure/temp/etc below instead of n/a

zxaber replied:
Do you mean the outside environment pressure/temp? That's a neat idea, but if I do that, it'll probably be in a different PR. Thanks for catching the typo.

zxaber commented on code/game/mecha/equipment/tools/work_tools.dm:
```diff
@@ -461,3 +461,50 @@
+	desc = "A pressurized canopy attachment kit for an Autonomous Power Loader Unit \"Ripley\" MK-I mecha, to convert it to the slower, but space-worthy MK-II design. This kit cannot be removed, once applied."
+	icon_state = "ripleyupgrade"
+
+/obj/item/mecha_parts/mecha_equipment/ripleyupgrade/can_attach(obj/mecha/working/ripley/M as obj)
```
I'll double-check it, but I was using the clamp procs as a template and that's how the clamp is done.

zxaber replied:
Yep, the `as obj` part is unnecessary and works removed. I'll update the PR.

zxaber commented on code/game/mecha/equipment/tools/work_tools.dm:
```diff
@@ -461,3 +461,50 @@
+		N.equipment += E
+		E.forceMove(N)
+		E.chassis = N
+		M.equipment -= E
```
I want to sorta force them over because it's supposed to seem like it's the same mech. As long as the new mech is updated correctly and the equipment's chassis pointer is updated, the equipment doesn't seem to care.

AnturK replied:
All it needs to break horribly is for equipment to use callbacks, this is just unsafe. If something depends on attachment order preserve that info then use it here.

zxaber replied:
Annoyingly enough, just running E.detach() and then E.attach(M) doesn't actually move anything over, so all the equipment gets deleted. If I'm going to need to get detach and attach put in here to get this merged, I'm gonna have to work at it a bit.

zxaber replied:
Alright, got it to work. I still need to remove the equipment manually from M, since I assume that's normally handled with the detach button functionality, but otherwise detaching and re-attaching works.

Turns out what I was worried about isn't a big deal, since it's the equipment's try_attach() function that actually handles whether or not an equipment is allowed to attach.

zxaber commented on code/game/mecha/equipment/tools/work_tools.dm:
```diff
@@ -461,3 +461,50 @@
+		if(istype(E, /obj/item/mecha_parts/concealed_weapon_bay)) //why is the bay not just a variable change who did this
+			E.forceMove(N)
+	N.obj_integrity = M.obj_integrity //This is not a repair tool
+	M.wreckage = 0
```
In the Destroy() proc, it skips making wreckage if the wreckage value is zero. Behind the scenes, I'm making a new mech and using qdel on the old one, but it's supposed to look like we're converting the old one, so I don't want to leave behind wreckage.

AnturK replied:
Oh i misread this is the original mecha, all good.

zxaber committed:
Some changes

DNA lock is kept during conversion
Maint_access setting is kept during conversion
Name is kept if it is not the default MK-I name
Typo 'enviroment' fixed

zxaber committed:
Requested changes

zxaber committed:
requested changes

zxaber commented:
So, one last thing I ran into;

If the MK-I is shot by a rocket, the damage is passed onto the occupant. An explosion is spawned on the target (inside the mech), which damages surrounding tiles but not the mech. The pilot gets wrecked, but it doesn't feel correct that the mech doesn't have a scratch on it.

I'm not really sure how to go about changing it. My first thought would be to give projectiles an `explosive` variable flag, and set it to true on all explosive projectiles, and just have those hit the mech instead, but I don't know if that's the correct fix. Any suggestions?

zxaber committed:
This is not my bug, but I will fix it

If the mech's radio is destroyed, it no longer run-times and breaks the view status popout.

ExcessiveUseOfCobblestone commented on code/game/mecha/equipment/tools/work_tools.dm:
```diff
@@ -461,3 +461,54 @@
+	if(M.occupant) //We're actualy making a new mech and swapping things over, it might get weird if players are involved
+		to_chat(loc, "<span class='warning'>[M] must be unoccupied before this conversion kit can be applied.</span>")
+		return 0
+	return 1
```
t/f these return values smh

zxaber replied:
I just copied what the existing equipment use ¯\\_(ツ)_/¯

ExcessiveUseOfCobblestone commented on code/game/mecha/equipment/tools/work_tools.dm:
```diff
@@ -461,3 +461,54 @@
+	if(!N)
+		return
+	qdel(N.cell)
+	N.cell = null
```
490/491 replace with `QDEL_NULL(N.cell)`

ExcessiveUseOfCobblestone commented on code/game/mecha/equipment/tools/work_tools.dm:
```diff
@@ -461,3 +461,54 @@
+		N.cell = M.cell
+		M.cell.forceMove(N)
+		M.cell = null
+	N.step_energy_drain = M.step_energy_drain //For the scanning module
```
lol does it delete the stock parts on construction? I hate mech code...

zxaber replied:
Imagine trying to figure this all out. It's a mess.

ExcessiveUseOfCobblestone commented on code/game/mecha/mecha.dm:
```diff
@@ -80,8 +80,15 @@
 	var/melee_cooldown = 10
 	var/melee_can_hit = 1
 
+	var/silicon_pilot = 0 //set to true if an AI or MMI is piloting.
```
is this a boo

zxaber replied:
fixed

ExcessiveUseOfCobblestone commented on code/game/mecha/mecha.dm:
```diff
@@ -80,8 +80,15 @@
+	var/silicon_pilot = 0 //set to true if an AI or MMI is piloting.
+
+	var/enter_delay = 40 //Time taken to enter the mech
+	var/enclosed = 1 //Set to zero for open-cockpit mechs
```
boo?

zxaber replied:
fixed

ExcessiveUseOfCobblestone commented on code/game/mecha/mecha.dm:
```diff
@@ -80,8 +80,15 @@
+
+	var/enter_delay = 40 //Time taken to enter the mech
+	var/enclosed = 1 //Set to zero for open-cockpit mechs
+	var/silicon_icon_state = "" //if the mech has a different icon when piloted by an AI or MMI
```
you should null this if it's suppose to be "empty"

ExcessiveUseOfCobblestone commented on code/game/mecha/mecha.dm:
```diff
@@ -127,7 +134,8 @@
 	add_radio()
 	add_cabin()
-	add_airtank()
+	if (enclosed == 1)
```
```suggestion
	if (enclosed)
```

yep it's a boo smh...

ExcessiveUseOfCobblestone commented on code/game/mecha/mecha.dm:
```diff
@@ -80,8 +80,15 @@
+
 	//Action datums
 	var/datum/action/innate/mecha/mech_eject/eject_action = new
+//	if (enclosed)
```
is this comment needed?

ExcessiveUseOfCobblestone commented on code/game/mecha/mecha.dm:
```diff
@@ -280,6 +288,10 @@
+	if(!enclosed && !silicon_pilot && occupant && occupant != user)
+		to_chat(user, "You can see [occupant] inside.")
+	if(!enclosed && silicon_pilot)
+		to_chat(user, "[src] appears to be piloting itself...")
```
```DM
if(!enclosed)
    if(silicon_pilot)
        to_chat(user, "[src] appears to be piloting itself...")
    else if(occupant && occupant != user) //!silicon_pilot implied
        to_chat(user, "You can see [occupant] inside.")

ExcessiveUseOfCobblestone commented on code/game/mecha/mecha.dm:
```diff
@@ -400,14 +412,19 @@
 	. = ..()
 	if(speaker == occupant)
-		if(radio.broadcasting)
+		if(radio && radio.broadcasting)
```
```suggestion
		if(radio?.broadcasting)
```

ExcessiveUseOfCobblestone commented on code/game/mecha/mecha.dm:
```diff
@@ -483,7 +500,7 @@
 	if(.)
 		events.fireEvent("onMove",get_turf(src))
-	if (internal_tank.disconnect()) // Something moved us and broke connection
+	if (internal_tank && internal_tank.disconnect()) // Something moved us and broke connection
```
```suggestion
	if (internal_tank?.disconnect()) // Something moved us and broke connection
```

ExcessiveUseOfCobblestone commented on code/game/mecha/mecha.dm:
```diff
@@ -509,7 +526,7 @@
 		to_chat(user, "<span class='notice'>You climb out from [src].</span>")
 		return 0
-	if(internal_tank.connected_port)
+	if(internal_tank && internal_tank.connected_port)
```
```suggestion
	if(internal_tank?.connected_port)
```

ExcessiveUseOfCobblestone commented on code/game/mecha/mecha.dm:
```diff
@@ -915,12 +937,16 @@
+	if (silicon_icon_state)
+		icon_state = silicon_icon_state
+	else
+		icon_state = initial(icon_state)
```
why is this check not just in update_icon?

zxaber replied:
I added the check everywhere the original code initial'd the icon_state variable. I suppose I could just check the status of silicon_pilot and silicon_icon_state at the top of update_icon.

ExcessiveUseOfCobblestone replied:
mech code...

zxaber replied:
Is it possible to check if an icon exists? If so, I could rename the empty icon to ripley-silicon, do an `if (silicon_pilot && <initial icon + "-silicon" exists>)`, and just remove the silicon_icon_state variable.

ExcessiveUseOfCobblestone replied:
no, besides the var is fine I'm just grossed out by the same check everywhere.

zxaber replied:
It's slightly better now.

ExcessiveUseOfCobblestone commented on code/game/mecha/working/ripley.dm:
```diff
@@ -16,6 +17,9 @@
 	var/hides = 0
+	enclosed = 0 //Normal ripley has an open cockpit design
+	enter_delay = 10 //can enter in a quarter of the time of other mechs
+	opacity = 0 //Ripley has a window
```
to see just the occupant lol

ExcessiveUseOfCobblestone commented on code/game/mecha/working/ripley.dm:
```diff
@@ -56,33 +60,68 @@
+/obj/mecha/working/ripley/check_for_internal_damage(list/possible_int_damage,ignore_threshold=null)
+	if (!enclosed)
+		possible_int_damage -= MECHA_INT_TEMP_CONTROL //if we don't even have an air tank, these two doesn't make a ton of sense.
+		possible_int_damage -= MECHA_INT_TANK_BREACH
```
`possible_int_damage -= (MECHA_INT_TEMP_CONTROL + MECHA_INT_TANK_BREACH)`
??

zxaber replied:
The internal tank breach is impossible if the MK-I doesn't have a tank at all, and the internal fire didn't make mush sense when you should be able to see the fire and put it out externally. This removes them from the pool of possible events.

ExcessiveUseOfCobblestone replied:
sorry for not being clear, my concern is that you're doing 2 expressions when it can be shortened to one.

zxaber replied:
Oh, gotcha. I'll make the change

ExcessiveUseOfCobblestone commented on code/game/mecha/working/ripley.dm:
```diff
@@ -56,33 +60,68 @@
+	wreckage = /obj/structure/mecha_wreckage/ripley/mkii
+	enclosed = 1
+	enter_delay = 40
+	silicon_icon_state = ""
```
```suggestion
	silicon_icon_state = null
```

ExcessiveUseOfCobblestone commented on code/game/mecha/working/ripley.dm:
```diff
@@ -56,33 +60,68 @@
+	step_in = 4
+	armor = list("melee" = 40, "bullet" = 20, "laser" = 10, "energy" = 20, "bomb" = 40, "bio" = 0, "rad" = 0, "fire" = 100, "acid" = 100)
+	wreckage = /obj/structure/mecha_wreckage/ripley/mkii
+	enclosed = 1
```
```suggestion
	enclosed = TRUE
```

ExcessiveUseOfCobblestone commented on code/game/mecha/working/ripley.dm:
```diff
@@ -56,33 +60,68 @@
+	enclosed = 1
+	enter_delay = 40
+	silicon_icon_state = ""
+	opacity = 1
```
```suggestion
	opacity = TRUE
```

ExcessiveUseOfCobblestone commented on code/game/mecha/working/ripley.dm:
```diff
@@ -56,33 +60,68 @@
 	armor = list("melee" = 40, "bullet" = 30, "laser" = 30, "energy" = 30, "bomb" = 60, "bio" = 0, "rad" = 0, "fire" = 100, "acid" = 100)
 	max_equip = 5 // More armor, less tools
 	wreckage = /obj/structure/mecha_wreckage/ripley/firefighter
+	enclosed = 1
```
```suggestion
	enclosed = TRUE
```

ExcessiveUseOfCobblestone commented on code/game/mecha/working/ripley.dm:
```diff
@@ -56,33 +60,68 @@
 	wreckage = /obj/structure/mecha_wreckage/ripley/firefighter
+	enclosed = 1
+	enter_delay = 40
+	silicon_icon_state = ""
```
```suggestion
	silicon_icon_state = null
```

ExcessiveUseOfCobblestone commented on code/game/mecha/working/ripley.dm:
```diff
@@ -56,33 +60,68 @@
+	enclosed = 1
+	enter_delay = 40
+	silicon_icon_state = ""
+	opacity = 1
```
```suggestion
	opacity = TRUE
```

ExcessiveUseOfCobblestone commented on code/game/mecha/working/ripley.dm:
```diff
@@ -56,33 +60,68 @@
+	enclosed = 1
+	enter_delay = 40
+	silicon_icon_state = ""
+	opacity = 1
```
see above, L121/3/4

Angustmeta commented:
Can smoke and foam affect the pilot?

zxaber committed:
Summary (required)

zxaber commented:
I _think_ so for the smoke, since the pilot will be on environment air, but I'm not certain. I'll have to test.

Not sure at all about the foam.

zxaber committed:
Indent-driven if statements are stupid

ExcessiveUseOfCobblestone commented:
they're practically copypastas of each other so probably.

zxaber committed:
Found the plus key

zxaber commented:
"Error generating diffs!"? What does that even mean?

zxaber committed:
Centered mech sprites

noticed the side view sprites were a pixel off.

zxaber committed:
update_icon changes

zxaber committed:
thing

AnturK commented:
Don't worry about it, diffbott having seizure.

davethwave commented:
Yay, can't wait to mass produce mechs with metal.

ExcessiveUseOfCobblestone commented:
try to abuse it so I can remove to my heart's content
